lib/lightning/auditing/audit.ex

```diff
@@ -118,7 +118,12 @@ defmodule Lightning.Auditing.Audit do
     field :item_id, Ecto.UUID
     embeds_one :changes, Changes
     field :actor_id, Ecto.UUID
-    field :actor, :map, virtual: true
+
+    field :actor_type, Ecto.Enum,
+      values: [:project_repo_connection, :trigger, :user]
+
+    field :actor_display_identifier, :string, virtual: true
+    field :actor_display_label, :string, virtual: true
```

Comment:
@jyeshe My initial intent was to retain the actor map, but when I started experimenting with unions, I ran into the following error:

```
** (Ecto.SubQueryError) the following exception happened when compiling a subquery.

         ** (Ecto.QueryError) atoms, structs, maps, lists, tuples and sources are not allowed as map values in subquery, got: `%{actor: &1}` in query:
         
         from a0 in Lightning.Auditing.Audit,
           prefix: "public",
           left_join: u1 in Lightning.Accounts.User,
           on: u1.id == a0.actor_id,
           where: a0.actor_type == "Lightning.Accounts.User",
           union_all: (from a0 in Lightning.Auditing.Audit,
           prefix: "public",
           left_join: p1 in Lightning.VersionControl.ProjectRepoConnection,
           on: p1.id == a0.actor_id,
           where: a0.actor_type == "Lightning.VersionControl.ProjectRepoConnection",
           select: merge(a0, %{actor: p1})),
           select: merge(a0, %{actor: u1})
```

There are ways to correct this and still retain the actor map but  I am not sure that the benefit outweighs the cost currently, as my understanding of the deliverable is that the primary intent is to ensure that the data is captured.

Therefore, my preference is to apply a less elegant and less flexible approach to satisfy the current view requirements so that I can maintain momentum on the capturing of audit events. I am more than willing to revisit this if (a) one of the new events demands it or (b) there is time to do so once the primary objectives have been met.

Reply:
Got it, in this case I would do just a single query with a `select: %{audit: a, user: u, trigger: t, provision: p}`:
```
query = from a in Lightning.Auditing.Audit, left_join: u in Lightning.Accounts.User, on: [id: a.actor_id], left_join: t in Lightning.Workflows.Trigger, on: [id: a.actor_id], select: %{audit: a, user: u, trigger: t}
```
It would allow keeping a generic model and use a single select on the SQL side.
To set the actor would be as clean as:
```
Enum.map(audits, &Map.put(&1.audit, :actor, &1.user || &1.trigger || &1.provision))
```

Reply:
@jyeshe My other feature branch got to the point where I needed to update the query so I thought I would just update it in this PR rather than delaying it.

I ended up with something very close to what you suggested, except I added an `actor_display` virtual element that provides an actor-agnostic interface that the view can use. Should the view's needs change, the `actor_display` can be cheaply modified to accommodate those needs (within reason :)) and likewise in the case where we add more actor structs.

Reply:
Understood the decision for `actor_display_for` but what was the reason on removing the generic actor field? WIth this field we would already have the type.
![image](https://github.com/user-attachments/assets/90064a67-fe27-49f4-a7a6-63c52bd15e0d)

So instead of having four fields, the select would already answer it with the type and the struct:
```
    field :actor_type, Ecto.Enum,
      values: [:project_repo_connection, :trigger, :user]

    field :project_repo_connection, :map, virtual: true
    field :trigger, :map, virtual: true
    field :user, :map, virtual: true
```
With the existing `actor` we wouldn't need this function:
```
 defp extract_actor_type(struct_name) do
    case struct_name do
      Lightning.VersionControl.ProjectRepoConnection -> :project_repo_connection
      Lightning.Workflows.Trigger -> :trigger
      Lightning.Accounts.User -> :user
    end
  end
```
and instead of having double checking types we would have only one:
```
  defp actor_display_for(%{
         trigger: %Trigger{type: type} = _actor,
         actor_type: :trigger
       }) do
    %{
      identifier: nil,
      label: Atom.to_string(type) |> String.capitalize()
    }
  end
```
becomes only:
```
  defp actor_display_for(%{actor: %Trigger{type: type}}) do
    %{
      identifier: nil,
      label: Atom.to_string(type) |> String.capitalize()
    }
  end
```

Reply:
Sounds good sir Rory! Thank you!

Reply:
Thank you @jyeshe